### PR TITLE
Improved L1->L2 fee calculation logic for Functions

### PR DESCRIPTION
**Pull request by KuphJr**
- Updates L1 fee calculation logic for Arbitrum to use correct calldata size
- Update L1 fee calculation logic for Optimism / Base to use new `GasPriceOracle.getL1FeeUpperBound(calldataSize)` function introduced in the Fjord hardfork

### CODE REVIEW

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (justinkaseman) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]:
```suggestion
  /// @dev ARB_L1_FEE_DATA_PADDING_SIZE is the L1 data padding for Arbitrum
```

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (justinkaseman) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol:42]: Add more words here explaining what "upper limit" means

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (justinkaseman) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: ** not ^

**Review by FelixFan1992 — DISMISSED** [review]
looks good to me.

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (matYang) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: 10^6 is returned in GasPriceOracle as `decimals()`, do we know if OP will always stick to 6 decimals? if not, probably better to read this from GasPriceOracle or have it configurable

Comment (matYang) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: the official formula is `16*base_fee_scalar*base_fee + blob_base_fee_scalar*blob_base_fee`, might be missing the `16*` here

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (matYang) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: out of curiosity, is this documented in OP docs?

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: Thank you for this comment! I will look into this further.

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: There is some nuance here that has to do with compression.
Technically they multiply (4 * zeroBytesCount + 16 * nonzeroBytesCount) / 16
However here we assume all data is uncompressed to get the "upper limit", therefore
(16 * nonzeroBytesCount) / 16 == nonzeroBytesCount
so we shouldn't need the 16*

Can you confirm that this makes sense @matYang ?

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: I have investigated further and I cannot find a source.
I believe that my confusion was due to conflation with Arbitrum.
Given the issues you have already noticed thus far, I am seeking external reviews from both the Arbitrum & OP/Base teams to confirm this implementation.
Thank you very much for your comments.

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (matYang) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: I mean the 16 is not on the entire gas price, it is on the base fee (calldata) portion and not on the blob portion right?

I get it that you want to assume every byte is non-zero and have `tx_compressed_size  = dataSizeBytes` to cap the worst case, then the formula becomes

```
l1_data_fee = dataSizeBytes * weighted_gas_price
= dataSizeBytes * ( (16*base_fee_scalar*base_fee) + blob_base_fee_scalar*blob_base_fee)
```

File: .changeset/gorgeous-garlics-rescue.md
Comment (justinkaseman) [inline .changeset/gorgeous-garlics-rescue.md]: Can we tack this tag on to the other changeset (neat-peas-reflect) so that there aren't two patch bumps?

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: I looked at the GasPriceOracle contract and the `decimals()` function is marked as "legacy".

Not sure if it is better to use a "legacy" function or just use the hardcoded value of 10^6. I think 10^6 should be safe as it is explicitly mentioned here: https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da

What do you think?

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: Note that we already know we will need to overhaul this logic again in order to support the new L1 fee calculation logic proposed for Fjord, so that might make the hardcoded 10^6 value not that big of a deal. I highly doubt this would change prior to the hardfork.

Context: https://chainlink-core.slack.com/archives/C028EC3TF70/p1715980777258499

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (matYang) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: regarding `(16 * 10 ** 6)`, the extra 16 is not needed here right? the weighted gas price is already gas per size_in_byte according to OP official doc

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (matYang) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: regarding (16 * 10 ** 6), the extra 16 is not needed here right? without it, the `weightedGasPrice` is already gas per byte according to OP official doc, if you are multiplying it with `dataSizeBytes`, I assume you do want it to be gas per byte

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: Alright, upon further research, I have learned the following:
- Optimism's docs say they use the _whole signed transaction size_ to determine the L1 fee
- Arbitrum's docs say they use the _calldata size_ + 140 "padding bytes"
However, I analysed past Functions transactions & determined that 140 can be thought of as an "upper limit" to the "static" transaction data (static tx data meaning data that is not contained in the calldata such as the signature, etc.)
On Optimism, it looks like the difference between the [calldata length](https://sepolia.basescan.org/tx/0x83b76a9abd8bf370e078ac71fe05d3a384fbf0aa176126dc632899b8e63703a3) and the [full raw tx length](https://sepolia.basescan.org/getRawTx?tx=0x83b76a9abd8bf370e078ac71fe05d3a384fbf0aa176126dc632899b8e63703a3) is 1874 - 1764 = 110.
In my latest commit, I am using the calldata.length+140 to determine the L1 fee for both networks.

File: contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol
Comment (ibrajer) [inline contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: I think you are missing division by 16? This is stated in the [GitHub link](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da) above and also [here](https://docs.optimism.io/stack/transactions/fees#ecotone)

Also, can you explain why the data padding size is 140 bytes? I'm looking at the [OracleGasPrice implementation details](https://optimistic.etherscan.io/address/0xb528d11cc114e026f138fe568744c6d45ce6da7a#code) and it looks like they are taking only [68 bytes padding](https://github.com/ethereum-optimism/optimism/blob/233ede59d16cb01bdd8e7ff662a153a4c3178bdd/packages/contracts-bedrock/contracts/L2/GasPriceOracle.sol#L110) for the transaction signature into account.

Comment (ibrajer) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: For example, this is my code based on the official OP GasPriceOracle implementation:
```
  function _calculateOptimismL1DataFee(uint256 calldataSizeBytes) internal view returns (uint256) {
    // 68 bytes of padding for transaction signature
    uint256 l1GasUsed = calldataSizeBytes * 16 + (68 * 16);
    uint256 scaledBaseFee = L1BLOCK.baseFeeScalar() * 16 * OVM_GASPRICEORACLE.l1BaseFee();
    uint256 scaledBlobBaseFee = L1BLOCK.blobBaseFeeScalar() * L1BLOCK.blobBaseFee();
    uint256 fee = l1GasUsed * (scaledBaseFee + scaledBlobBaseFee);
    return fee / (16 * 10 ** OVM_GASPRICEORACLE.decimals());
  }
```

Comment (matYang) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: you can cancel out the 16 from the equation:
```
return fee / (16 * 10 ** OVM_GASPRICEORACLE.decimals());
===
return l1GasUsed * (scaledBaseFee + scaledBlobBaseFee) / (16 * 10 ** OVM_GASPRICEORACLE.decimals());
===
return (calldataSizeBytes * 16 + (68 * 16)) * (scaledBaseFee + scaledBlobBaseFee) / (16 * 10 ** OVM_GASPRICEORACLE.decimals());
===
return (calldataSizeBytes + 68) * (scaledBaseFee + scaledBlobBaseFee) / 10 ** OVM_GASPRICEORACLE.decimals());
```

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: I actually think it is correct.
We start with this:
`(zeroes*4 + ones*16) * (16*l1BaseFee*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar) / 16e6`
We then assume all bytes are non-zero
`(ones*16) * (16*l1BaseFee*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar) / 16e6`
We can then cancel out
`ones * (16*l1BaseFee*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar) / 1e6`

Does this make sense? Please double-check carefully as I may be incorrect.

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: The padding is something I have found to be a "max upper limit" to the `tx calldata size` - `full raw tx size`
However, I want to come up with a better justification and back it up with more documentation. I will be asking the OP/Base team.

At least the Arbitrum team specifically documented the 140 number. OP docs just say "size of the full signed tx"

Comment (ibrajer) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: > Does this make sense? Please double-check carefully as I may be incorrect.

@KuphJr I checked and it makes sense 👍 @matYang explained it very well in the comment above
I didn't get to the point of calculating the correct padding size yet for VRF fulfillment, but I do want to run some tests and make comparisons. That's why I was interested in understanding how 140 bytes was selected as the padding size.
If the OP team gives you any further clarification on this topic, do let me know!

Comment (roberto-bayardo) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: if calldataSizeBytes is the unsigned transaction length then you'll want to add 68 bytes for the signature, not 140.

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol]: Just updated the PR!
I think that should be the final commit needed.

**Review by justinkaseman — DISMISSED** [review]
Approving, pending staging sniff tests should be done before merging.

**Review by justinkaseman — APPROVED** [review]

**Comment by KuphJr** [comment]
Changes were ported over to this PR: https://github.com/smartcontractkit/chainlink/pull/13815

**Pull request closed without merging**
